src/redisearch_rs/buffer/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    /// Returns the remaining capacity of the buffer (capacity - length).
+    pub fn remaining_capacity(&self) -> usize {
+        self.capacity - self.len
+    }
+
+    /// Advance the buffer by `n` bytes.
+    ///
+    /// This increases the buffer's length by `n` bytes, effectively marking more of the buffer as
+    /// "initialized" without actually writing any data. Typically used after directly writing to
+    /// the buffer's memory.
+    ///
+    /// # Safety
+    ///
+    /// * `n` must not exceed the remaining capacity of the buffer.
+    /// * The new bytes added by this call must be initialized.
+    pub unsafe fn advance(&mut self, n: usize) {
+        debug_assert!(n <= self.remaining_capacity());
+        self.len += n;
+    }
+}
+
+/// Redefines the `BufferWriter` struct from `buffer.h`
+///
+/// A writer for appending data to a `Buffer`. Provides a mechanism to write bytes to a `Buffer`
+/// while tracking the current write position with a cursor.
+///
+/// # Safety Invariants
+///
+/// * `buf` must point to a valid, properly initialized `Buffer`.
+/// * `cursor` must point to a valid position within the Buffer's allocated memory.
+/// * The `BufferWriter` does not own the `Buffer`, only borrows it.
+/// * If the buffer is grown (implicitly) as part of a write operation, all previously obtained
+///   pointers into the buffer are invalidated.
+///
+/// # Usage
+///
+/// It's best to use this through the `std::io::Write` implementation, which handles
+/// buffer growth and cursor position management automatically.
+///
+/// ```no_run
+/// # use std::io::Write;
+/// # use buffer::{Buffer, BufferWriter};
+/// # use std::ptr::NonNull;
+/// let buffer_ptr: NonNull<Buffer> = todo!();
+/// let mut writer = unsafe { BufferWriter::for_buffer(buffer_ptr) };
+///
+/// // Write data to the buffer
+/// writer.write(b"Hello, world!").unwrap();
+/// ```
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct BufferWriter {
+    pub buf: NonNull<Buffer>,
+    pub cursor: NonNull<u8>,
+}
+
+impl BufferWriter {
+    /// Create a new `BufferWriter` for the given buffer.
+    ///
+    /// # Safety
+    ///
+    /// We assume `buf` is a valid pointer to a properly initialized `Buffer`.
+    pub unsafe fn for_buffer(buffer: NonNull<Buffer>) -> Self {
+        Self {
+            buf: buffer,
+            // Safety: We assume `buf` is a valid pointer to a properly initialized `Buffer`.
+            cursor: unsafe { buffer.as_ref() }.data,
+        }
+    }
+}
+
+impl std::io::Write for BufferWriter {
+    fn write(&mut self, bytes: &[u8]) -> std::io::Result<usize> {
+        // Safety: We assume `buf` is a valid pointer to a properly initialized Buffer.
+        let buffer = unsafe { self.buf.as_mut() };
+
+        // Check if we need to grow the buffer to accommodate the new data
+        if buffer.len + bytes.len() > buffer.capacity {
```

Comment:
maybe use checked addition here too, to avoid nasty wrapping bugs if (for whatever reason) people call this will really large slices?

Reply:
I'm not sure if 32-bit systems are targetted by Redis and for 64-bit machines, this will be one hell of a huge buffer that is extremely unlikely to happen IMO. Perhaps one of Redis folks can chime in here? @kei-nan ?

Reply:
I can certainly add a `debug_assert` here though.